File: c/collatz.c
/* find a path to one
    if even, divide by 2
    if odd, multiply by 3 and add 1
    return number of steps to 1.
    */
#include <stdio.h>
char isEven(long int n);

long int collatz(long int n)
{
    long int count = 0;

    while(n>1)
    {
        if(isEven(n))
        {
            n/=2;
            count++;
        }
        else
        {
            n=n*3 +1;
            count++;
        }

    }
    /* added check to ensure this really did settle on n equals 1 */
    /* this caught an error where collatz(113383) after count = 120 had gone negative...
    all results are likely invalid beyond this point */
    if(n==1) return count;
    else return -1;
}
char isEven(long int n)
{
    return ((n%2)==0);
}

File: python/collatz.py
from isEven import isEven


def collatz(n):
    count = 0

    while(n>1):
        if(isEven(n)):
            n//=2
            count +=1
        else:
            n=n*3+1
            count +=1
    if(n==1):
        return count
    else:
        return -1

def problem14(upper):
    """identify starting point of longest collatz path below uppper"""
    n=1
    max = 0
    maxn = 0
    while(n<upper):
        c=collatz(n)
        if c>max:
            max=c
            maxn=n
            print(max,maxn)
        n+=1
    print("max", max, "n", maxn)

